File: Programmers/DFS/다단계 칫솔 판매.py
def dfs(profitMap, parentsMap, seller, profit):
    if seller == '-' or profit < 1: return

    share = profit // 10
    pro = profit - share
    profitMap[seller] += pro
    dfs(profitMap, parentsMap, parentsMap[seller], share)


def solution(enroll, referral, seller, amount):
    answer = []
    n = len(enroll)
    m = len(seller)
    profitMap = {}
    parentsMap = {}

    for i in range(n):
        profitMap.setdefault(enroll[i], 0)
        parentsMap.setdefault(enroll[i], "-")
        parentsMap[enroll[i]] = referral[i]

    for i in range(m):
        dfs(profitMap, parentsMap, seller[i], amount[i] * 100)

    for name in enroll:
        answer.append(profitMap[name])

    return answer

File: Programmers/DFS/다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

map<string, string> parentsMap;
map<string, int> profitMap;

void DFS(string seller, int profit) {
    if (seller == "-" || profit < 1) return;

    int share = profit / 10;
    int tmp = profit - share;
    profitMap[seller] += tmp;
    DFS(parentsMap[seller], share);
}

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> answer;

    for (int i = 0; i < enroll.size(); i++) {
        parentsMap[enroll[i]] = referral[i];
    }

    for (int i = 0; i < seller.size(); i++) {
        int total = amount[i] * 100;
        DFS(seller[i], total);
    }

    for (auto name : enroll) {
        answer.push_back(profitMap[name]);
    }


    return answer;
}